### DISCUSSION TITLE
Performance of substr()

### DISCUSSION THREAD

**Discussion by rvjansen** [adesutherland/CREXX#104] [Progress Updates]
So after this saga of bugs I finally could do what I set out to do: measure the performance gains of the new concchar and bct instructions. I does not vary wildly. A slight surprise is that the old version is somehow minimally faster:
```
Performance counter stats for 'rxvm testsubstr oldsubstr length':

              0.40 msec task-clock                #    0.645 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
                68      page-faults               #    0.169 M/sec
           1532362      cycles                    #    3.798 GHz
           1114322      instructions              #    0.73  insn per cycle
            227965      branches                  #  564.951 M/sec
              9238      branch-misses             #    4.05% of all branches

       0.000625429 seconds time elapsed

       0.000701000 seconds user
       0.000000000 seconds sys
```
for the old version, and
```
 Performance counter stats for 'rxvm testsubstr substr length':

              0.43 msec task-clock                #    0.643 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
                68      page-faults               #    0.159 M/sec
           1633521      cycles                    #    3.811 GHz
           1123368      instructions              #    0.69  insn per cycle
            230152      branches                  #  536.907 M/sec
              9204      branch-misses             #    4.00% of all branches

       0.000667093 seconds time elapsed

       0.000728000 seconds user
       0.000000000 seconds sys
```
for the new version. This on x86_64. For (a much slower, and busier (but that really does not change the cycles) ARM64 system:
```
 Performance counter stats for 'rxvm testsubstr oldsubstr length':

              2.93 msec task-clock                #    0.552 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
                44      page-faults               #    0.015 M/sec
           2625773      cycles                    #    0.896 GHz
           2244974      instructions              #    0.85  insn per cycle
   <not supported>      branches
             15489      branch-misses

       0.005309360 seconds time elapsed

       0.000000000 seconds user
       0.005087000 seconds sys

```
for the old version without concchar and bct, and 
```
 Performance counter stats for 'rxvm testsubstr substr length':

              2.94 msec task-clock                #    0.556 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
                46      page-faults               #    0.016 M/sec
           2624058      cycles                    #    0.893 GHz
           2252553      instructions              #    0.86  insn per cycle
   <not supported>      branches
             15416      branch-misses

       0.005286341 seconds time elapsed

       0.000000000 seconds user
       0.005130000 seconds sys
```
for new. So the performance gain is really very small. The programming however, is much more entertaining.

**Comment by rvjansen** [comment]
I up-arrowed these for a number of times now. It is too close to call.

**Reply by Peter-Jacob** [reply]
statistically spoken: the differences are not significant.

**Comment by adesutherland** [comment]
I am surprised but pleased because it kind of tells us (I think) that the threaded architecture really helps the branch prediction and takes a lot of overhead from our RXAS instruction calls. Quite remarkable really

**Comment by rvjansen** [comment]
So the question is, why is the difference so small, because it is evident that the new version executes less rxas instructions. I think the answer is here, that we spend a lot more time in the unicode functions (if we look at the execution profile):
```
  31.78%  rxvm                             [.] run
  15.40%  rxvm                             [.] utf8nlen
  11.77%  rxvm                             [.] utf8codepointcalcsize
   8.52%  rxvm                             [.] string_set_byte_pos
   7.27%  rxvm                             [.] string_concat_char
   4.26%  rxvm                             [.] utf8catcodepoint
   3.44%  rxvm                             [.] utf8codepoint
   2.60%  rxvm                             [.] extend_string_buffer
   1.70%  rxvm                             [.] prep_string_buffer
   1.63%  rxvm                             [.] set_string
   1.26%  rxvm                             [.] utf8codepointsize
   1.18%  rxvm                             [.] set_const_string
   1.11%  rxvm                             [.] set_int
   1.06%  [kernel]                         [k] 0xffffffffa58e3227
```

**Reply by Peter-Jacob** [reply]
These are bad news. In my understanding one char can occupy 1 to 4 bytes, what if we add an attribute to each string containing 1 or M, saying it is a simple string with one byte notation, or multi byte. In my testing my test strings were all 1-byte. I assume if I had used German special chars they were muli bytes.

**Reply by rvjansen** [reply]
note that the profile is from some runs of mid.rxas, with the loop around it; otherwise I cannot get good percentages for the instruction mix. The other numbers are from a one-time execution of the current test case.

**Comment by rvjansen** [comment]
It is nowadays the reality. I mean, oorexx still says that length(‘René’) is 5. I also suggested a bit that marks a string as >1 byte at the moment it is cheapest to determine, so searches can switch implementations, but Java copes with it and other languages also do. It is funny how many banks and company websites you can break by putting in a 🙂

**Comment by adesutherland** [comment]
No No! UTF is UTF, and mixing and matching that way you suggest Pater
would be a path to hell! ... :-)

I think we instead need to step back and review each instruction and each
BIF to ensure that they are doing UTF efficiently.

1. For example we have utf8nlen() top of the resource list. Well this
required a full string scan so is indeed a nightmare. However there is no
reason why it is being used because the string length is stored and if you
concat a character (codepoint) the string length is just incremented ...;

2. Not sure of utf8codepointcalcsize() - I am surprised it is so slow or
either used so much, but perhaps that is inevitable - or perhaps it is
being used by utf8nlen() which explains it.

3. Finally can't we copy more than one character at a time ....

So Peter / Rene - my challenge is - why are we using utf8nlen() so much -
avoid!

A

See rxvmvars/h in interpreter - string_chars

struct value {
/* bit field to store value status - these are explicitly set (not
automatic at all) */
value_type status;
/* Value */
rxinteger int_value;
double float_value;
void *decimal_value; /* TODO */
char *string_value;
size_t string_length;
size_t string_buffer_length;
size_t string_pos;
#ifndef NUTF8
size_t string_chars;
size_t string_char_pos;
#endif
void *object_value;
/*
* Each value can either be owned by a stack frame or a variable pool
* The owner is responsible for freeing the value, but because value can
* be linked other stack frames and variable pools it is important that
* none of these free its memory when they are being freed themselves.
* The owner member is only used by parents to see if they are the real
* parent - if you like a paternity test!
* This also allows a value to be adopted by another parent (e.g. for a
* returned register from a procedure)
*/
void *owner;
};

**Reply by Peter-Jacob** [reply]
I did a quick scan of the instructions, I can't find any utf8nlen() function call, so I guess they wre called by using set_string, which is also not used explicitly by the instructions. So it seems we call them indirectly.  I can't guarantee the completeness of my searches, as I don't have a running environment on my laptop, but if true we must improve set_string and colleagues, which shouldn't be too hard as they are central functions.

**Reply by adesutherland** [reply]
I think I have found at least one issue

    load r1,"blah blah"

uses utfbnlen() and so we might be seeing the test harness overhead (rather than the test itself).

Anyway it is bad - I will make a change so that the assembler (at build time) uses the utf8nlen() to get the character length of the constant string and store this in the constant pool. Then the runtime load will be much faster as it wont need to do string scan!

I will make this change in devlop and merge into f0030

**Comment by adesutherland** [comment]
set_string() uses utf8nlen() because it takes a c string - do you use
set_string()?

Perhaps you can use a version of

static void string_concat_char(value *v1, value *v2);

In this it appends an integer from v2 - or you could easily make a version
called

string_concat_int(value *v1, int v2);

Or something?

What do you guys think?

Adrian

**Comment by rvjansen** [comment]
I have a bit of good news: it is not so bad! I compared the same test case on brexx, regina and oorexx.
The first is cREXX:
```
➜  tests git:(feature/f0030) ✗ perf stat rxvm ../rxas/testsubstr ../rxas/substr ../rxas/length
Test of Substr function
567890
67890.
bc
bc
bc..
ab
bcéééé

 Performance counter stats for 'rxvm ../rxas/testsubstr ../rxas/substr ../rxas/length':

              0.48 msec task-clock                #    0.662 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
                69      page-faults               #    0.143 M/sec
           1837400      cycles                    #    3.819 GHz
           1196414      instructions              #    0.65  insn per cycle
            243624      branches                  #  506.355 M/sec
              9504      branch-misses             #    3.90% of all branches

       0.000726824 seconds time elapsed

       0.000786000 seconds user
       0.000000000 seconds sys
```
Regina comes in second:
```
➜  tests git:(feature/f0030) ✗ perf stat rexx tstsubstr.rexx
567890
67890.
bc
bc
bc..
ab
    10 +++ say substr('abcdefgh',2,3,'é')
Error 40 running "/home/rvjansen/apps/CREXX/lib/rxfns/tests/tstsubstr.rexx", line 10: Incorrect call to routine
Error 40.23: SUBSTR argument 4 must be a single character; found "é"

 Performance counter stats for 'rexx tstsubstr.rexx':

              0.70 msec task-clock                #    0.726 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
               224      page-faults               #    0.322 M/sec
           2642293      cycles                    #    3.801 GHz
           1862766      instructions              #    0.70  insn per cycle
            362439      branches                  #  521.407 M/sec
             11414      branch-misses             #    3.15% of all branches

       0.000957477 seconds time elapsed

       0.000996000 seconds user
       0.000000000 seconds sys
```

BREXX is a bit slower:

```
➜  tests git:(feature/f0030) ✗ perf stat brexx tstsubstr.rexx
sh: 1: OPTIONS: not found
     2 *-* options levelb
       +++ RC(32512) +++
567890
67890.
bc
bc
bc..
ab
    10 *-* say substr('abcdefgh',2,3,'é')
Error 40 running tstsubstr.rexx, line 10: Incorrect call to routine

 Performance counter stats for 'brexx tstsubstr.rexx':

              1.31 msec task-clock                #    0.809 CPUs utilized
                 2      context-switches          #    0.002 M/sec
                 0      cpu-migrations            #    0.000 K/sec
               178      page-faults               #    0.136 M/sec
           5118482      cycles                    #    3.920 GHz
           4007367      instructions              #    0.78  insn per cycle
            786357      branches                  #  602.298 M/sec
             27511      branch-misses             #    3.50% of all branches

       0.001613497 seconds time elapsed

       0.001704000 seconds user
       0.000000000 seconds sys
```
ooRexx is a *lot* slower:
```
➜  tests git:(feature/f0030) ✗ perf stat oorexx tstsubstr.rexx
567890
67890.
bc
bc
bc..
ab
    10 *-* say substr('abcdefgh',2,3,'é')
Error 40 running /home/rvjansen/apps/CREXX/lib/rxfns/tests/tstsubstr.rexx line 10:  Incorrect call to routine.
Error 40.23:  SUBSTR argument 4 must be a single character; found "é".

 Performance counter stats for 'oorexx tstsubstr.rexx':

              4.70 msec task-clock                #    0.923 CPUs utilized
                 3      context-switches          #    0.638 K/sec
                 0      cpu-migrations            #    0.000 K/sec
               837      page-faults               #    0.178 M/sec
          18078438      cycles                    #    3.843 GHz
          16314575      instructions              #    0.90  insn per cycle
           3106856      branches                  #  660.350 M/sec
             70394      branch-misses             #    2.27% of all branches

       0.005099387 seconds time elapsed

       0.005074000 seconds user
       0.000000000 seconds sys
```

Note that none of the other rexxen gets the unicode pad right: they complain about a character not being a character, which would be confusing to young people.

So with being the only one to get it right, and being the fastest of them all, we should not really worry.
I also think this validates the approach of writing rxas for functions, or even levelb rexx (after we measure that, of course).

**Comment by rvjansen** [comment]
on the other hand, if we can improve on the current state - have at it!

**Comment by rvjansen** [comment]
one interesting tidbit: Rexx370 on z/VM aces this test:
```
Ready; T=0.01/0.01 16:57:41 
 say substr('abcde',2,8,'é')
bcdeéééé                    
```

**Comment by adesutherland** [comment]
That is surprising!

**Comment by rvjansen** [comment]
yes, sorry, corrected that quickly to z/VM when I saw what I typed

**Comment by rvjansen** [comment]
but fortunately, the Rexx on VM370 also does this without problem!

**Reply by adesutherland** [reply]
I am bored with documentation! May I join the performance competition and do a substr in rexx? I know it's duplication of effort ....

**Reply by Peter-Jacob** [reply]
Certainly, remember, the winner pays the pints!

**Comment by rvjansen** [comment]
Of course, we need that; I don’t have time today because of social obligations. So great that you mention it beforehand!

René.

**Comment by adesutherland** [comment]
This is my first version - but I want to do some microcode and compiler optimisations before the beer order. 

Still I would be interested in any performance results ...

```
/* Classic REXX Runtime Library */

options levelb /* Written in REXX Level B */

/* Tests */
say substr("René Vincent Jansen",1,4,".")
say substr("René Vincent Jansen",6,7,"")
say substr("12345678",5,6,"é")
say substr("12345678",10,6,"é")
say substr("12345678",10,6,"éé")

/* Raise() Internal Function to Raise a runtime error */
raise: procedure = .int
  arg type .string, code .string, parm1 .string
  /* TODO Variable number of arguments */
  say "RUNTIME" type "ERROR:" code parm1
  /*
  exit /* TODO - the compiler does not supports exit yet */
  */
  return

/* Length() Procedure */
length: procedure = .int
  arg string1 .string /* Pass by reference */
  result = 0
  assembler strlen result,string1
  return result

/* Substr() Procedure */
substr: procedure = .string
  arg string1 .string, start=.int, length=.int, pad .string
  /* TODO length and pad are optional */

  padchar = 0 /* Is an integer */
  char = 0    /* Again needs to be integer */
  output = ''
  inputLength = 0;
  padLength = 0;

  /* Check Start */
  if start < 1 then call raise "syntax", "40.??", start /* Invalid start */
  start = start - 1 /* 1 base to zero base */

  /* Check length of pad */
  assembler strlen padLength,pad;
  if padLength > 1 then call raise "syntax", "40.23", pad /* Invalid pad length */

  /* Get the Length of the input string */
  assembler strlen inputLength,string1
  inputLength = inputLength - start;


  if inputLength > 0 then do
    /* Yes there are characters needed from string1 */
    if length <= inputLength then do
      /* Just copy from string1 - no padding needed */
      do i = start to start + length
         assembler do
           strchar char,string1,i
           appendchar output,char
         end
      end
    end
    else do
      /* Copy all of string1 and then pad */
      do i = start to start + inputLength - 1
         assembler do
           strchar char,string1,i
           appendchar output,char
         end
      end

      /* Then add pads */
      /* Get Pad Char as integer */
      if padLength = 0 then pad = " "
      assembler strchar padchar,pad,padchar /* padchar is set to 0 so can use it as char pos */
      /* Append the pads */
      do i = 1 to length - inputLength
        assembler appendchar output,padchar
      end
    end
  end

  else do
    /* Nothing from string 1 - just Pad */
    /* Get Pad Char as integer */
    if padLength = 0 then pad = " "
    assembler strchar padchar,pad,padchar /* padchar is set to 0 so can use it as char pos */
    /* Append pads */
    do i = 1 to length
      assembler appendchar output,padchar
    end
  end

  /* Done */
  return output
```

**Reply by Peter-Jacob** [reply]
you know that there is  CONCCHAR which combines both instructions, don't you? 
assembler do
           strchar char,string1,i
           appendchar output,char
end

results in:
assembler concchar output,string1,i

**Reply by adesutherland** [reply]
Ah - I had forgotten! Thanks

**Reply by rvjansen** [reply]
tried to compile this but compilers from both develop and f0030 ended with a segfault:
```
➜  zero git:(feature/f0030) ✗ lldb ~/apps/crexx_debug/compiler/rxc zerosubstr
(lldb) target create "/Users/rvjansen/apps/crexx_debug/compiler/rxc"
Current executable set to '/Users/rvjansen/apps/crexx_debug/compiler/rxc' (x86_64).
(lldb) settings set -- target.run-args  "zerosubstr"
(lldb) run
Process 22418 launched: '/Users/rvjansen/apps/crexx_debug/compiler/rxc' (x86_64)
Process 22418 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
    frame #0: 0x0000000100061059 rxc`step4_walker(direction=out, node=0x0000000100404c20, payload=0x00007ffeefbfeaf0) at rxcpbval.c:652:48
   649 	                    n1->target_type = n2->value_type;
   650 	                    if (n2->child->node_type == VAR_REFERENCE) {
   651 	                        n1->is_ref_arg = 1;
-> 652 	                        n1->symbol->writeUsage = 1;
   653 	                        if (!n1->symbol) mknd_err(n1, "CANT_BE_PASSED_BY_REF");
   654 	                    }
   655 	                    n1 = n1->sibling;
Target 0: (rxc) stopped.
(lldb)
```

**Reply by rvjansen** [reply]
same on Linux X86_64:
```
Program received signal SIGSEGV, Segmentation fault.
step4_walker (direction=out, node=0x5555555db000, payload=0x7fffffffdcf0) at /home/rvjansen/apps/CREXX/compiler/rxcpbval.c:652
652	                        n1->symbol->writeUsage = 1;
```

**Reply by adesutherland** [reply]
Sorry that wasted some of your life! I fixed that yesterday but forgot to commit - I have pushed fix into f0030 and included the global.rexx zero lib

**Comment by adesutherland** [comment]
And BTW the first slightly complex program compiled by rxc!

![image](https://user-images.githubusercontent.com/4367482/132963031-7320a303-240b-47c3-8dfd-8e47ff4b8fc9.png)

**Reply by Peter-Jacob** [reply]
Adrian, can you please mail me the assembly?

**Reply by rvjansen** [reply]
yes maybe if you post the assembly I can run the profile

**Reply by adesutherland** [reply]
Sorry - I have also a small fix for the compiler. I am undecided - should I push this for f0033 or develop? Can we / are we going to merge in f0033 to develop in the next very few days?

Attached files as a zip

[global.zip](https://github.com/adesutherland/CREXX/files/7149435/global.zip)

**Reply by rvjansen** [reply]
Erm, if you mean f0030 then we can merge at any moment; better merge early than late.

**Comment by adesutherland** [comment]
So the fixes / improvements I want to do before the "final" results are declared (interim results will be motivational!)

1. The compiler is "leaking" return registers (see in global.rxas how the say in the tests each have a higher register)

2. improve load r1,"string" so that the string length calculation is done by the assembler at build time

3. End of loops are generally like
```
   * Line 54: {Implicit "BY 1"}
   inc r8
   * Line 54: {DO-END}
   br l151
``` 
So we should add incbr / decbr super instructions and change the compiler (or assembler - see below).

4. Beginning of loops are like 
```
l151:
   * Line 54: to start + length
   igt r0,r8,r7
   brt l145,r0
```
So we need super instructions like (Peter did you do something like this?)

    ibrgt l145,r8,r7

This is good enough for the common "do i = 1 to max" In time we may need the whole family (ibrlte, ibteq, fbrlt etc etc) - lots :-(

And this is a good time to try the do this by changing the *assembler* to have a peephole optimiser step.

What is interesting is there are two kinds of RXAS instructions - those optimised for hand written assembler, and those friendly for a less flexible compiler. In designing real CPUs I wonder how many people see this dichotomy?

**Reply by rvjansen** [reply]
John Cocke was very firm on including support for compilers in hardware instead of making it easy on the programmer. Consequently, RS/6000 and Power assembler is very hard to handcode, compared to s/370.

**Reply by Peter-Jacob** [reply]
The do i=n to m compilation is my only (slight) concern it uses too many instructions,  whereas BCT/BCTNM does it all with one instruction. But I understand it can't be translated from a Rexx statement.

**Reply by Peter-Jacob** [reply]
I am happy to add your new instructions, but it has to wait for another week, I am returning on 24. fom Italy

**Reply by adesutherland** [reply]
Thanks Peter - and I wish I was in Italy(!)

**Comment by rvjansen** [comment]
very promising, I changed the tests to do the same, except for filling out the optionals:
```
➜  zero git:(feature/f0030) ✗ perf stat rxvm zerosubstr
567890
67890.
bc4
bc
bc..
abc
bcde

 Performance counter stats for 'rxvm zerosubstr':

              0.43 msec task-clock                #    0.649 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
                69      page-faults               #    0.161 M/sec
           1628536      cycles                    #    3.811 GHz
           1153098      instructions              #    0.71  insn per cycle
            233244      branches                  #  545.789 M/sec
              8932      branch-misses             #    3.83% of all branches

       0.000657989 seconds time elapsed

       0.000727000 seconds user
       0.000000000 seconds sys
```

**Reply by adesutherland** [reply]
I will try and improve the load instruction now - this might allow us to see differenced between the versions more clearly.

**Comment by rvjansen** [comment]
global itself does more or less the same:
```
➜  zero git:(feature/f0030) ✗ perf stat rxvm global
René
Vincent
5678éé
éééééé
RUNTIME syntax ERROR: 40.23 éé
éééééé

 Performance counter stats for 'rxvm global':

              0.43 msec task-clock                #    0.640 CPUs utilized
                 0      context-switches          #    0.000 K/sec
                 0      cpu-migrations            #    0.000 K/sec
                67      page-faults               #    0.155 M/sec
           1650805      cycles                    #    3.820 GHz
           1121747      instructions              #    0.68  insn per cycle
            227267      branches                  #  525.833 M/sec
              9179      branch-misses             #    4.04% of all branches

       0.000674963 seconds time elapsed

       0.000744000 seconds user
       0.000000000 seconds sys

```

**Comment by rvjansen** [comment]
so, I think this validates a decision to build them all in Rexx level b and only descend when needed

**Reply by adesutherland** [reply]
I agree - it feels like there is definitely room for improvement with the compiler and instruction set (and looking at the profiler, some good hints of where to attack). The load instruction just needs fixing for example and there are a few super instructions that will help loops. And of course we need an instruction to copy more than 1 character at a time.

But the point for me is that already performance is good - we are not suffering a catastrophe using REXX for library functions. It is better than good enough in my opinion.

Another point - there is an overhead calling a function. From the profiler, the frame_f (factory), free_frame, free_value and value_int_f are all associated with calling a function, the new stack frame and so on. It will be possible (some time in the perhaps distant future) to get the rxc compiler to inline REXX functions called from REXX but it will not be possible to inline RXAS or native functions - this alone might eat up any other advantage from a RXAS or native implementation.

**Reply by rvjansen** [reply]
Unless, of course, that native implementation can do its own loop unrolling, or, e.g. vectorised processing

**Comment by rvjansen** [comment]
profile when running this in a large loop, with say changed into an 'x=':
``` 
  31.92%  rxvm                             [.] run
   7.40%  libc-2.31.so                     [.] _int_free
   6.84%  libc-2.31.so                     [.] __libc_calloc
   5.98%  libc-2.31.so                     [.] _int_malloc
   4.82%  rxvm                             [.] free_value
   4.26%  rxvm                             [.] frame_f
   3.66%  rxvm                             [.] value_int_f
   3.34%  rxvm                             [.] set_int
   2.95%  rxvm                             [.] string_concat_char
   2.94%  rxvm                             [.] free_frame
   2.79%  rxvm                             [.] utf8nlen
   2.44%  rxvm                             [.] string_set_byte_pos
   2.18%  libc-2.31.so                     [.] __memset_avx2_unaligned_erms
   1.93%  rxvm                             [.] utf8codepointcalcsize
   1.88%  rxvm                             [.] prep_string_buffer
   1.85%  libc-2.31.so                     [.] cfree@GLIBC_2.2.5
   1.64%  rxvm                             [.] copy_value
   1.40%  rxvm                             [.] utf8codepoint
   1.28%  rxvm                             [.] set_string
   1.20%  rxvm                             [.] utf8catcodepoint
   1.00%  [kernel]                         [k] 0xffffffffa58e3227
   0.93%  libc-2.31.so                     [.] __memmove_avx_unaligned_erms
   0.91%  rxvm                             [.] extend_string_buffer
   0.75%  rxvm                             [.] set_const_string
   0.47%  rxvm                             [.] utf8codepointsize
```

**Reply by Peter-Jacob** [reply]
So we have a winner: the winner pays the toll!

**Reply by adesutherland** [reply]
Well ... we shall see (early days!) :-)

**Reply by adesutherland** [reply]
Anyway this adds to Rene's mission to get us physically to the same place (at the same time) ...

**Comment by Peter-Jacob** [comment]
It seems I have got my CREXX environment up and running. Please let me know if we remain on f0030. I'll try to knock up the new instructions if I get the time ;)

**Comment by adesutherland** [comment]
Well done - f0030 is fine :-)

**Comment by rvjansen** [comment]
with all the good news, there is also a slight dissonant: two testcases fail when compared to the rxas version:
```
say substr('abc',2,2,' ')
```
should say 'bc', actual 'bc4'
```
say substr('abcdefgh',1,2,'.')
```
should say 'ab', actual 'abc'

**Comment by adesutherland** [comment]
>  with all the good news, there is also a slight dissonant
The universe must stay balanced!

**Comment by adesutherland** [comment]
Pushed a fix (I can't count from 0!)

**Comment by rvjansen** [comment]
the fix works. Q: is the name global.rexx an indication that we are going to package all the rexx bifs into one file? If that is the case, we could work on the rest of them, maybe employing an '#include' scheme for the moment, until we have interfile linking on demand for rexx running. And then still - for performance reading less, somewhat larger files, is probably better than doing file I/O on lots of small files.

**Reply by adesutherland** [reply]
Sorry for the delay in replying; this is just a short term solution until we have a linker / library tool that can create a library. 

But in principle we could put many BIFS here until it gets too big to be manageable (and it will be a good stress test for rxc - I have no proof that it will compile really big files!). We can organise them later into different source files, and if you want to use the c pre-processor as a short term fix - well why not? 

The idea is that in the future, rxvm will automatically load something like "rexxlvlc" library, or "rexxlvlg" and so on as determined by a flag in the rxbin file. Or the linker tool will link the rexxlvlc library to the rxbin and then rxvm will not need to link at run time. Or both as options.

As part of the PoC we could make rxvm automatically include global.rxbin to save typing? What do you think? On the other hand, it might be better to keep it manual so we know exactly what is happening ...

In terms of performance (large vs lots of small files) - not sure but it would only impact building and linking, not runtime.

**Comment by rvjansen** [comment]
I would like to see the standard parts of Rexx included as soon as possible; it enables us to write execs with a growing amount of bifs without a lot of ceremony. From global (but I'd rather we'll call it rexfns - or rename the directory around it to 'global', so there is correspondence with the libraries and the subdirectories of lib - at least for crexx deliverables).

I am interested in the plans for the linkage editor. After working all these years with the OS linkage editor/binder, I could see the refreshing approach of VM that just concats text files - I assume the start command or a loader fixes up the addresses.

On the other hand, a really versatile linker would be good, so we can choose how to layer things using the linker, like including modules from different libraries. The OS  linker can do even more, like punching in exits.

For me, the inclusion of optional function arguments is the highest priority at the moment, for testing the BIF's.

**Reply by adesutherland** [reply]
> For me, the inclusion of optional function arguments is the highest priority at the moment, for testing the BIF's.

Noted - I'll get on to it!

**Comment by rvjansen** [comment]
I am going to: integrate the current test cases for the current lib/rxas library into one suite
After that, I am going to ready the bif testcases. The missing ones from rexfns/zero can be gotten from lib/rxas; I think afterwards we will use the fastest variant. I am also going to check in some prototype native functions, for playing around with.

**Comment by rvjansen** [comment]
About halfway through this, but I have a line in f0030: lib/rxas/testbifs.rxas (I generate that from .mxas but I checked this in for easy reference) that is kryptronite to the assembler. It is the line say "Test of Left function, no pad", although it can be something else also, like say "hello" at that point. The error is:
```
cpp -E testbifs.mxas | awk '/^[^#]/{ print }' > testbifs.rxas
rxas testbifs
Assertion failed: (yyruleno<(int)(sizeof(yyRuleName)/sizeof(yyRuleName[0]))), function yy_reduce, file /Users/rvjansen/apps/CREXX/assembler/rxasgrmr.c, line 1100.
make: *** [testbifs.rxas] Abort trap: 6
```
Some other observations: my substr is still no good, but thanks to the testing harness this becomes clear now.
And, the testing harness cannot count.

**Reply by rvjansen** [reply]
when moving things around, it looks like some maximum is reached

**Reply by rvjansen** [reply]
now on Linux X86_64, same file but other errors:
cpp -E testbifs.mxas | awk '/^[^#]/{ print }' > testbifs.rxas
rxas testbifs

Error Severity 1
232:4 - Error at "load", invalid header directive
233:4 - Error at "load", invalid header directive
234:4 - Error at "load", invalid header directive
235:4 - Error at "call", invalid header directive
236:4 - Error at "load", invalid header directive
237:4 - Error at "ieq", invalid header directive
238:4 - Error at "call", invalid header directive
239:4 - Error at "call", invalid header directive
240:4 - Error at "exit", invalid header directive
241:1 - Error at "", totally confused after headers
corrupted size vs. prev_size
make: *** [makefile:6: testbifs.rxas] Aborted (core dumped)

and the debug version:
```
Starting program: /home/rvjansen/apps/crexx_debug/assembler/rxas testbifs
corrupted size vs. prev_size

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) where
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7de7859 in __GI_abort () at abort.c:79
#2  0x00007ffff7e523ee in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7f7c285 "%s\n")
    at ../sysdeps/posix/libc_fatal.c:155
#3  0x00007ffff7e5a47c in malloc_printerr (str=str@entry=0x7ffff7f7a43a "corrupted size vs. prev_size") at malloc.c:5347
#4  0x00007ffff7e5aaeb in unlink_chunk (p=p@entry=0x55555558ac80, av=0x7ffff7fadb80 <main_arena>) at malloc.c:1454
#5  0x00007ffff7e5c00b in _int_free (av=0x7ffff7fadb80 <main_arena>, p=0x555555581030, have_lock=<optimized out>)
    at malloc.c:4342
#6  0x0000555555557db6 in main (argc=2, argv=0x7fffffffded8) at /home/rvjansen/apps/CREXX/assembler/rxasmain.c:265
(gdb)
```

**Comment by adesutherland** [comment]
OK ... Just looking at it now

**Comment by adesutherland** [comment]
Well I am not getting the same error - but I am getting memory issues. This is a good thing as it might mean you are just seeing a symptom of memory corruption in my code (the alternative is something weird in the lemon parser generator - which might be hard to fix!)

**Comment by adesutherland** [comment]
Well - my laziness!

rxasmain.c line 190 

```
   scanner.binary.binary = malloc(sizeof(bin_code) * 5000); /* todo */
   scanner.binary.const_pool = malloc(sizeof(unsigned char) * 5000); /* TODO */
```

The  programme binary and constant  pool size is limited to 5000 instructions / bytes! 

If you change it to 50000 I think it "fixes" the problem - But I will work on and push a change to dynamically grow both of these which will cause a merge conflict when you pull it.

Who know I would have a program > 5k!

PS - I did mark them as todo - but todo's are always for tomorrow ....

**Reply by adesutherland** [reply]
In fact I will push the 50000 "fix" now and the real fix later - that way the merge conflict will be avoided

**Reply by rvjansen** [reply]
The temp fix works great, and I could resume work immediately. This must be the largest program until now. I am glad it was an artificial limit and not an indication of some problem somewhere.
